File: workspace/2023/2023-01-23/run-length-encoding.python/run_length_encoding_test.py
import unittest
from run_length_encoding import encode, decode

TEST_DATA_TUPLES = [
    # [xs, runs, encoded_runs]
    [[],                             [],                                   []],
    [['a'],                          [['a']],                              [('a', 1)]],
    [['a', 'a'],                     [['a', 'a']],                         [('a', 2)]],
    [['a', 'a', 'a'],                [['a', 'a', 'a']],                    [('a', 3)]],
    [['a', 'b'],                     [['a'], ['b']],                       [('a', 1), ('b', 1)]],
    [['a', 'b', 'b'],                [['a'], ['b', 'b']],                  [('a', 1), ('b', 2)]],
    [['a', 'a', 'b', 'c', 'c', 'c'], [['a', 'a'], ['b'], ['c', 'c', 'c']], [('a', 2), ('b', 1), ('c', 3)]],
]

class RunLengthEncodingTestCase(unittest.TestCase):

    def test_encode_decode(self):
        for i, (xs, runs, encoded_runs) in enumerate(TEST_DATA_TUPLES):
            self.assertEqual(xs, decode(encode(xs)))

    def test_encode(self):
        for i, (xs, runs, encoded_runs) in enumerate(TEST_DATA_TUPLES):
            self.assertEqual(encoded_runs, encode(xs))

    def test_decode(self):
        for i, (xs, runs, encoded_runs) in enumerate(TEST_DATA_TUPLES):
            self.assertEqual(xs, decode(encoded_runs))

if __name__ == "__main__":
    unittest.main()

File: workspace/2023/2023-02-19/run-length-encoding.ruby/run_length_encoding_test.rb
require 'test/unit'
require 'run_length_encoding'

class TestCase < Test::Unit::TestCase

    TEST_DATA_LIST = [
        # [xs, encoded_runs]
        [''.chars,       []],
        ['a'.chars,      [['a', 1]]],
        ['aa'.chars,     [['a', 2]]],
        ['ab'.chars,     [['a', 1], ['b', 1]]],
        ['aabccc'.chars, [['a', 2], ['b', 1], ['c', 3]]],
    ]

    TEST_DATA_LIST.each_with_index do |(xs, encoded_runs), i|
        define_method("test_encode_decode_#{i + 1}") do
            assert_equal(xs, decode(encode(xs)))
        end
    end

    TEST_DATA_LIST.each_with_index do |(xs, encoded_runs), i|
        define_method("test_encode_#{i + 1}") do
            assert_equal(encoded_runs, encode(xs))
        end
    end

    TEST_DATA_LIST.each_with_index do |(xs, encoded_runs), i|
        define_method("test_decode_#{i + 1}") do
            assert_equal(xs, decode(encoded_runs))
        end
    end

    def test_xs_to_runs
        assert_equal(
            [],
            xs_to_runs(''.chars))
        assert_equal(
            [['a', 'a'], ['b'], ['c', 'c', 'c']],
            xs_to_runs('aabccc'.chars))
    end

end

